services/api/service.go

```diff
@@ -381,6 +398,47 @@ func (api *RelayAPI) startValidatorRegistrationDBProcessor() {
 	}
 }
 
+// simulateBlock sends a request for a block simulation to blockSimRateLimiter.
+func (api *RelayAPI) simulateBlock(opts blockSimOptions) error {
+	t := time.Now()
+	simErr := api.blockSimRateLimiter.send(opts.ctx, opts.req, opts.highPrio)
+	if simErr != nil {
+		log := opts.log.WithField("simErr", simErr.Error())
+		log.WithError(simErr).WithFields(logrus.Fields{
+			"duration":   time.Since(t).Seconds(),
+			"numWaiting": api.blockSimRateLimiter.currentCounter(),
+		}).Info("block validation failed")
+	} else {
+		opts.log.WithFields(logrus.Fields{
+			"duration":   time.Since(t).Seconds(),
+			"numWaiting": api.blockSimRateLimiter.currentCounter(),
+		}).Info("block validation successful")
+	}
+	return simErr
+}
+
+// startOptimisticBlockProcessor keeps listening on the channel and validating incoming blocks asynchronously.
+func (api *RelayAPI) startOptimisticBlockProcessor() {
+	for opts := range api.optimisticBlockC {
+		err := api.simulateBlock(opts)
+		if err != nil {
```

Comment:
If simulation fails but the bid is not the winning there would be no trace in the DB, right? May be worth recording demotions in the DB (instead of refunds). Demotion row would have enough information to make refunds, when appropriate.

Reply:
Yep, exactly. Demotions should never happen so if they do happen it's best to have the full context to be able to debug what happened. I expect this will also simplify the code :)

Reply:
sounds good! resolving this comment because the other thread we have on this has more content :-)